text/2978-stack_vec.md

```diff
@@ -0,0 +1,185 @@
+- Feature Name: `stack_vec`
+- Start Date: 2020-09-27
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes the creation of a new "growable" vector named `StackVec` that manages stack memory and can be seen as an alternative for the built-in structure that handles heap-allocated memory, aka `alloc::vec::Vec<T>`.
+
+# Motivation
+[motivation]: #motivation
+
+`core::collections::StackVec<T>` has several use-cases and should be conveniently added into the standard library due to its importance.
+
+### Unification
+
+There are a lot of different crates about the subject that tries to do roughly the same thing, a centralized implementation would stop the current fragmentation.
+
+### Optimization
+
+Stack-based allocation is generally faster than heap-based allocation and can be used as an optimization in places that otherwise would have to call an allocator. Some resource-constrained embedded devices can also benefit from it.
+
+### Building block
+
+Just like `Vec`, `StackVec` is a primitive vector where high-level structures can use it as a building block. For example, a stack-based matrix or binary heap.
+
+### Useful in the real world
+
+`arrayvec` is one of the most downloaded project of `crates.io` and is used by thousand of projects, including Rustc itself.
+
+### Compiler internals
+
+Unstable features can be used internally to make operations more efficient, e.g., `specialization`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Instead of relying on a heap-allocator, stack-based memory area is added and removed on-demand in a last-in-first-out (LIFO) order according to the calling workflow of a program. Let's illustrate an imaginary function `super()` that allocates 32 bits before calling another function named `sub()` that also allocates 32 bits of auxiliary memory:
+
+```txt
+0 bits           32 bits          64 bits   
+
+
+|       |   ->   |       |   ->   |       |
+---------        ---------        ---------
+                 | super |        |  sub  |
+                 ---------        ---------
+                                  | super |
+                                  ---------
+```
+
+* Now `super()` returns and the exactly same thing happens in reverse order, dropping everything when out of scope.
+
+```txt
+64 bits          32 bits          0 bits   
+
+
+|       |   ->   |       |   ->   |       |
+---------        ---------        ---------
+|  sub  |        | super |
+---------        ---------
+| super |
+---------
+```
+
+`StackVec` takes advantage of this predictable behavior to reserve an exactly amount of uninitialized bytes up-front and these bytes form a buffer where elements can be included dynamically.
```

Comment:
I don't think the stack needs to be explained in an RFC or in the standard library documentation. The implementation of `StackVec` also would not really care about how or where it is allocated, so this section doesn't seem very relevant either.